Sources/Datadog/Core/Upload/DataUploadConditions.swift

```diff
@@ -8,6 +8,30 @@ import Foundation
 
 /// Tells if data upload can be performed based on given system conditions.
 internal struct DataUploadConditions {
+    enum Blocker {
+        case battery(level: Int, state: BatteryStatus.State)
+        case lowPowerModeOn
+        case networkReachability(description: String)
+    }
+
+    enum Report {
+        case go
+        case noGo(blockers: [Blocker])
+
+        static func && (lhs: Report, rhs: Report) -> Report {
+            switch (lhs, rhs) {
+            case (.go, .go):
+                return .go
+            case (.go, .noGo):
+                return rhs
+            case (.noGo, .go):
+                return lhs
+            case (let .noGo(blockers: lbs), let .noGo(blockers: rbs)):
+                return .noGo(blockers: lbs + rbs)
+            }
+        }
```

Comment:
what would `.noGo(blockers: [])` mean?

i think `go` is just lack of blockers, it's not a real state and this `&&` operator is just appending blockers.
if we change to
```swift
// before
func canUploadWith(...) -> Report { ... }
// after
func blockersForUpload(with: ...) -> [Blocker] { ... }
```
then `blockers.isEmpty` means `.go` state, we can remove `&&` operator, we don't need `switch case .noGo(...)` to unwrap blockers

wdyt?

Reply:
Yeah I have considered just returning the list of blockers and letting the call-site branch on the count. In the end, I thought that it would be more explicit to provide some `.go`/`.noGo` cases, but I overlooked the potential confusion if the `blockers` array was empty (although it's a case that should not happen, it would be a bug in `DataUploadConditions`).
I think your suggestion aligns with @ncreated's naming idea, that should work and make things simpler indeed. I'm going to correct that!